wagtail/admin/templates/wagtailadmin/home.html

```diff
@@ -6,17 +6,8 @@
     {{ block.super }}
     {{ media.css }}
 {% endblock %}
-
 {% block content %}
-    <header class="header merged header--home">
-        <div class="avatar"><img src="{% avatar_url user %}" alt="" /></div>
-
-        <div class="sm:w-ml-4">
-            <h1 class="header__title">{% block branding_welcome %}{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}{% endblock %}</h1>
-            <div class="user-name">{{ user|user_display_name }}</div>
-        </div>
-    </header>
-
+    {% include "wagtailadmin/shared/header.html" with title=site_name avatar=1 username=user|user_display_name merged=1 %}
```

Comment:
Also the title is wrong - it should be the result of 
`{% block branding_welcome %}{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}{% endblock %}`

But - you may need to set up this translation at the view level.

Reply:
ok, did a quick test and this should do the trick.

In the file ...`wagtail/admin/views/home.py`

```

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        panels = self.get_panels()
        site_details = self.get_site_details()

        context["media"] = self.get_media(panels)
        context["panels"] = sorted(panels, key=lambda p: p.order)
        context["user"] = self.request.user
        context["welcome_title"] = gettext_lazy("Welcome to the %(site_name)s Wagtail CMS") % { 'site_name': site_details['site_name'] } # add this line - it is the same translation block but done at the view level instead.

        return {**context, **site_details}
```

Reply:
Bear in mind that people routinely replace this text via template overrides - if we're moving this out of home.html, we'll need to provide an alternative way to customise it (and cover it in the release notes).

Reply:
Thanks Matt.
---
Do you know of a Django way to make a block become a value?

For example;
```
{% block branding_welcome as my_header_title %}{% endblock %}
```

Otherwise, not sure how we can make this easily overridable, but I will have a think;

* template inheritance (although, the current goal is to avoid that where possible as per parent issue)
* document that the home view can be extended, developers can customise the `get_context_data` and override the url with their custom home view (not sure if Django makes that easy though as the url name will exist twice).

Reply:
Really not sure where to head with this. I personally feel that inheritance here would be simpler and have more flexibility.

We could add a title block around the existing template and then use that to have the block behaviour. Developers will need to override the home page specific header template instead of the  home view. 

This will make it easier to align with how the ModelAdmin header template is built also. 

---

Another idea - what if we just put the welcome label in it's own template file? Or even don't create one and in our view's context we search for one and if not found fall back to the view default welcome.

---

I am concerned that if we try to force everything into component style template tags we may create something that is overall less flexible for developers. However, we do need to try to get the html content to be mostly consistent so that it is easier to do styling.  

But maybe it's not worth changing this? Just apply more styles on top for each version. 

@thibaudcolas please provide any additional ideas. 
@PaarthAgarwal have a read of all this and please add any thoughts you have.

Reply:
After some testing, I think the isolated welcome title template may be the simplest and supports the original flexibility, but will be a change for developers.

This 'pulls up' the content generation into the template view still instead of pushing down with layers of component/endcomponent templates.

---

Note: I tried a new `Component` but it felt like overkill for this.

## 1.  - create a new template file 

_Note: We could even preserve the block here so that developers have even less to change._

`wagtail/admin/templates/wagtailadmin/home/branding_welcome_title.html`
```html
{% load i18n %}
{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}
```

## 2. create a new method on the `HomeView` class

```python
    def get_welcome_title(self, context):
        return get_template("wagtailadmin/home/branding_welcome_title.html").render(context)
```

## 3. use this method in the `get_context_data`

```python
    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        panels = self.get_panels()
        site_details = self.get_site_details()

        context["media"] = self.get_media(panels)
        context["panels"] = sorted(panels, key=lambda p: p.order)
        context["user"] = self.request.user

        context = {**context, **site_details} # set up the context earlier

        context["welcome_title"] = self.get_welcome_title(context) # add this line

        return context
```

## 4. update documentation

Note: This assumes we will not be reusing the block `branding_welcome`...

````markdown

### `branding_welcome`

```{versionchanged} 4.0
```

To replace the welcome message on the admin home page (dashboard), create a template file `dashboard/templates/wagtailadmin/home/branding_welcome_title.html` that overrides the content:

```html+django
Welcome to Frank's Site
```
````

5. Update the release notes with the change.

* ... can be done when merged.

Reply:
I think it’s time to introduce new template tags, so we don’t have to switch back and forth between Python and templates to do relatively basic composition of components. Clearly the vanilla `include` and `block` aren’t fit for purpose when componentising a UI like this – we need a way to pass template fragments to arbitrary components via composition rather than inheritance.

My preference would be to introduce generic tags that can support those types of patterns, rather than specialised tags for each component. At the moment I think the flexibility of the [`fragment` tag from Slippers](https://mitchel.me/slippers/docs/template-tags-filters/#fragment) would be the best fit for Wagtail. Assuming it supports rendering blocks, it’d look something like:

```html
{% fragment as site_name %}
  {% block branding_welcome %}{% blocktrans trimmed %}Welcome to the {{ site_name }} Wagtail CMS{% endblocktrans %}{% endblock %}
{% endfragment %}
{% include "wagtailadmin/shared/header.html" with title=site_name avatar=1 username=user|user_display_name merged=1 %}
```

See for example:

- https://mitchel.me/slippers/docs/template-tags-filters/#fragment

Reply:
I agree with @thibaudcolas - if the situation calls for being able to render a component by composing template fragments, we should adopt a solution that supports that, rather than trying to work around that requirement.

Having said that - if moving the title element to its own include is the simplest way to unblock this, then let's do that. While it is a breaking change for site implementers, I don't think there was an expectation that we'd keep that block in place in that particular template forever, and this change should be compatible with whatever mechanism we introduce in future.

Reply:
Thanks all. This is super great feedback. I'll walk through this on my Monday call with Paarth. 

If it's ok - we will probably opt for Matt's approval to unblock this with the simplest way forward (own include for title element). 

I agree that the fragment concept and also the ongoing discussion about a component type template tag is super important. 

We may end up writing this again in a few weeks (depending on the capacity for the team to also bring in these new template tag systems) which is fine. In the short term for 4.0.RC looming, we are hoping for some quicker visual wins. 

I understand we need to be careful to consider existing use cases for developers though and will be more vigilant in assessing existing block usage.